storage/imx6ull-flash/flashdev.c

```diff
@@ -0,0 +1,434 @@
+#define ECC0_BITFLIP_STRENGHT 16
+
+
+/* Auxiliary functions */
+
+/* Function derived from the previous code phoenix-rtos-filesystems/jffs/phoenix-rtos/mtd.c */
+static int _flashmtd_errRead(int ret, flashdrv_meta_t *meta, int status, void *data)
+{
+	int max_bitflip = 0;
+	unsigned *blk;
+	int i, j, err = 0;
+	/* check current read status. if it is EBADMSG we have nothing to do here since
+	 * read is marked as uncorrectable */
+	if (status == -EBADMSG) {
+		return status;
+	}
+
+	/* check status for every block */
+	for (i = 0; i < sizeof(meta->errors) / sizeof(meta->errors[0]); i++) {
+		/* no error or all ones */
+		if (meta->errors[i] == 0x0 || meta->errors[i] == 0xff) {
+			continue;
+		}
+		/* uncorrectable errors but maybe block is erased and we can still use it */
+		/* TODO: check if this actually works */
+		else if (meta->errors[i] == 0xfe) {
+			err = 0;
+			if (i == 0) {
+				blk = (unsigned *)meta->metadata;
+				for (j = 0; j < sizeof(meta->metadata) / sizeof(unsigned); ++j) {
+					err += (sizeof(unsigned) * 8) - __builtin_popcount(*blk + j);
+				}
+
+				if (err > ECC0_BITFLIP_STRENGHT) {
+					return -EBADMSG;
+				}
+
+				memset(blk, 0xff, sizeof(meta->metadata));
+				max_bitflip = max(max_bitflip, err);
+			}
+			else {
+				blk = data + ((i - 1) * 512);
+				for (j = 0; j < 128; j++) {
+					err += (sizeof(unsigned) * 8) - __builtin_popcount(*(blk + j));
+				}
+
+				if (err > ECCN_BITFLIP_STRENGHT) {
+					return -EBADMSG;
+				}
+
+				memset(blk, 0xff, 512);
+				max_bitflip = max(max_bitflip, err);
+			}
+		}
+		/* correctable errors */
+		else {
+			max_bitflip = max(max_bitflip, (int)meta->errors[i]);
+		}
+	}
+
+	return max_bitflip >= ECC_BITFLIP_THRESHOLD ? -EUCLEAN : status;
+}
+
+
+/* MTD interface */
+
+static int flashmtd_erase(struct _storage_t *strg, off_t offs, size_t size)
+{
+	int res;
+	blkcnt_t pageID;
+	blkcnt_t eb, startBlock, endBlock, erased = 0; /* values in erase block units */
+	unsigned int pagesPerBlock;
+
+	if ((offs % strg->dev->mtd->erasesz) != 0 || (size % strg->dev->mtd->erasesz) != 0) {
+		return -EINVAL;
+	}
+
+	startBlock = offs / strg->dev->mtd->erasesz;
+	endBlock = (offs + size) / strg->dev->mtd->erasesz;
+	pagesPerBlock = strg->dev->mtd->erasesz / strg->dev->mtd->writesz;
+
+	mutexLock(strg->dev->ctx->lock);
+	for (eb = startBlock; eb < endBlock; ++eb) {
+		pageID = eb * pagesPerBlock;
+
+		res = flashdrv_isbad(strg->dev->ctx->dma, pageID);
+		if (res != 0) {
+			printf("flashmtd_erase: skipping bad block: %lld\n", eb);
+			continue;
+		}
+
+		res = flashdrv_erase(strg->dev->ctx->dma, pageID);
+		if (res < 0) {
+			printf("flashmtd_erase: error while erasing block: %lld - marking as badblock", eb);
+			res = flashdrv_markbad(strg->dev->ctx->dma, pageID);
+			if (res < 0) {
+				mutexUnlock(strg->dev->ctx->lock);
+				return res; /* can't mark as badblock, urecoverable error */
+			}
+		}
+		else {
+			++erased;
+		}
+	}
+	mutexUnlock(strg->dev->ctx->lock);
+
+	return erased;
```

Comment:
Why `erased` is type `long long` if You use it in this place as type `int`?
The same aplies to `pageID`.
`paigeID` is type `long long` but functions are declared as:
```
extern int flashdrv_isbad(flashdrv_dma_t *dma, uint32_t paddr);
extern int flashdrv_markbad(flashdrv_dma_t *dma, uint32_t paddr);
```
the second argument is type `uint32_t`.

If `offs` is to be of `long long` type, code where the divide 64bit is usedt can be improved by reducing the number of calls to `__aeabi_ldivmod()`.

Reply:
Changed to uint32_t